utxo/utxo.go

```diff
@@ -1296,37 +1334,77 @@ func getGasLimitFromTx(tx *pb.Transaction) (int64, error) {
 
 // verifyTxRWSets verify tx read sets and write sets
 func (uv *UtxoVM) verifyTxRWSets(tx *pb.Transaction) (bool, error) {
-	req := tx.GetContractRequest()
+	req := tx.GetContractRequests()
+	reservedRequests, err := uv.getReservedContractRequests(nil, tx)
+	if err != nil {
+		uv.xlog.Error("getReservedContractRequests error", "error", err.Error())
+		return false, err
+	}
+
+	if !verifyReservedContractRequests(reservedRequests, req) {
+		uv.xlog.Error("verifyReservedContractRequests error", "reservedRequests", reservedRequests, "req", req)
+		return false, fmt.Errorf("verify reservedContracts error")
+	}
+
 	if req == nil {
 		if tx.GetTxInputsExt() != nil || tx.GetTxOutputsExt() != nil {
 			uv.xlog.Error("verifyTxRWSets error", "error", ErrInvalidTxExt.Error())
 			return false, ErrInvalidTxExt
 		}
 		return true, nil
 	}
-	moduleName := req.GetModuleName()
-	vm, err := uv.vmMgr3.GetVM(moduleName)
+
+	env, err := uv.model3.PrepareEnv(tx)
 	if err != nil {
 		return false, err
 	}
-	env, err := uv.model3.PrepareEnv(tx)
+
+	contextConfig, err := contract.NewContextConfig(env.GetModelCache(), tx.GetInitiator(), tx.GetAuthRequire(), "", int64(0))
 	if err != nil {
 		return false, err
 	}
+
 	gasLimit, err := getGasLimitFromTx(tx)
 	if err != nil {
 		return false, err
 	}
+	gasRemain := int64(contract.MaxGasLimit)
 	uv.xlog.Trace("get gas limit from tx", "gasLimit", gasLimit, "txid", hex.EncodeToString(tx.Txid))
-	ctx, err := vm.NewContext(req.GetContractName(), env.GetModelCache(), gasLimit)
-	if err != nil {
-		return false, err
-	}
-	_, err = ctx.Invoke(req.MethodName, req.Args)
-	defer ctx.Release()
-	if err != nil {
-		return false, err
+
+	for i := 0; i < len(tx.GetContractRequests()); i++ {
+		if i == len(reservedRequests) {
+			gasRemain = gasLimit
+		}
+		tmpReq := tx.GetContractRequests()[i]
+		moduleName := tmpReq.GetModuleName()
+		vm, err := uv.vmMgr3.GetVM(moduleName)
+		if err != nil {
+			return false, err
+		}
+
+		contextConfig.GasLimit = gasRemain
+		contextConfig.ContractName = tmpReq.GetContractName()
+		ctx, err := vm.NewContext(contextConfig)
+		if err != nil {
+			// FIXME zq @icexin: need to return contract not found
+			uv.xlog.Error("verifyTxRWSets NewContext error", "err", err, "contractName", tmpReq.GetContractName())
+			if i < len(reservedRequests) && err.Error() == "leveldb: not found" {
+				continue
+			}
+			return false, err
+		}
+
+		_, err = ctx.Invoke(tmpReq.MethodName, tmpReq.Args)
+		if err != nil {
+			ctx.Release()
```

Comment:
when err == nil , ctx should also be released.

Reply:
ok~